README.md
# LwLL API

## Downloading Data

A CLI has been built for easy access to datasets that have been wrapped. From the root of this directory, you can use the CLI for the following pieces of functionality:

Before using the cli you want to make sure you have the required dependencies.

```sh
virtualenv -p python3.6 env
source env/bin/activate
pip install -r cli_requirements.txt
```

Using `conda`

```sh
conda create -n lwll python=3.6
conda activate lwll
pip install -r cli_requirements.txt
```

### List Available Datasets

-   `list_data` or `list`

List the available datasets that are live and ready for model training and development. Optionally, use a query to search for a specific string.

#### Ex. List data in the live lwll-dataset bucket

```sh
python download.py list_data
```

Example output:

```output
cifar100
coco2014
face_detection
imagenet_1k
mnist
voc2009
```

#### Ex. Search for cifar datasets

```sh
python download.py list cifar
```

Example output:

```output
cifar100
```

### Download Data

-   `download_data` or `download`

Download and unpacks datasets matching the query string. If the `external` stage is specified, labels are downloaded as well.

The `dataset` argument can be set to a single dataset name, such as `mnist` and `cifar100`; use `list_data` or `list` for the list of available datasets.

The `stage` argument can be either `development` or `external`, specifying which dataset variant to download for your current stage of development. Available shorthands: `d`, `dev`, `develop`, `e`, `ext`

The `output` parameter is used to specify the directory to put the datasets. It's value defaults to the current directory (`.`) if not specified.

The `check` parameter is a boolean flag set to False by default. If passed as argument, the script will check and list which datasets need to be updated without downloading them. 

You can also optionally download ALL of the datasets for a particular stage by passing in `ALL` to the `dataset` argument. This is helpful if you want to do something like daily checks that you have all of the available datasets and skip if you already have the datasets. In this case you could do something like:

```sh
python download.py download_data --dataset ALL --stage development --output ~/lwll_datasets 
python download.py download_data --dataset ALL --stage external --output ~/lwll_datasets
python download.py download_data --dataset ALL --stage external --output ~/lwll_datasets --check
```

#### Ex. Download CIFAR-100 Explicitly

Download the cifar100 dataset explicitly in the `development` stage:

```sh
python download.py download_data --dataset cifar100 --stage development --output ~/lwll_datasets
```

Example output:

```output
live/datasets/cifar100.tar.gz 119537664/146285881 (81.5359070777309%)
live/datasets/cifar100.tar.gz 146285881/146285881 (100.0%)
Extracting tarball...
Cleaning up...
Finished downloading: "cifar" to
        "~/lwll_datasets/development"
```

#### Ex. Download CIFAR-100 with Labels

Download the newest version of cifar100 dataset, force overwriting any version you may have in that location.

```sh
python download.py download_data --dataset cifar100 --stage external --output ~/lwll_datasets 
```

Example output:

```output
live/datasets/cifar100.tar.gz 146285881/146285881 (100.0%)
Extracting tarball...
Cleaning up...
Downloading labels...
live/labels/cifar100/full/labels_test.feather 232776/232776 (100.0%)
live/labels/cifar100/full/labels_train.feather 1162776/1162776 (100.0%)
live/labels/cifar100/sample/labels_test.feather 23464/23464 (100.0%)
live/labels/cifar100/sample/labels_train.feather 116368/116368 (100.0%)
Finished downloading: "cifar" to
        "~/lwll_datasets/external"
```

### File Structure

Upon downloading a dataset, it will be extracted within the output directory (ie. `~/lwll_datasets`) you've specified. Remember, labels aren't downloaded for the `development` stage.

```output
-- ~/lwll_datsets
-- -- external
-- -- -- cifar100
-- -- -- -- cifar100_full
-- -- -- -- -- test
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
-- -- -- -- -- train
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
-- -- -- -- cifar100_sample
-- -- -- -- -- test
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
-- -- -- -- -- train
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
-- -- -- -- labels_full
-- -- -- -- -- labels_test.feather
-- -- -- -- -- labels_train.feather
-- -- -- -- labels_sample
-- -- -- -- -- labels_test.feather
-- -- -- -- -- labels_train.feather
-- -- development
-- -- -- cifar100
-- -- -- -- cifar100_full
-- -- -- -- -- test
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
-- -- -- -- -- train
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
-- -- -- -- cifar100_sample
-- -- -- -- -- test
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
-- -- -- -- -- train
-- -- -- -- -- -- img*.*
-- -- -- -- -- -- etc.
```

## Machine Translation Monolingual Corpora

For machine translation tasks, monolingual corpora are often very useful. Since these monolingual datasets do not fit 
into one of our defined buckets of `development` or `external` because of the non uniform schema, we allow you to download 
the monolingual corpora in a similar fashion calling a different utility script. When downloading, you should place these 
to the same place you are downloading your other datasets, however, these will be placed into a sub directory called 
`monolingual_corpora`.

Currently there are currently 2 different corpora types, wikipedia dumps and Huma-Num OSCAR dumps. The wikipedia data 
contains 2 languages English and Arabic. Each wikipedia download is a single file. The OSCAR data includes 4 languages:
Sinhala, English, Polish, and Farsi. These corpora will download into a folder containing multiple text files. 

You can download these individually like so:

```sh
python download_monolingual.py download_data --dataset wiki-ar --output ~/lwll_datasets 
```

The two available options for `--dataset` for wikipedia include `wiki-ar` and `wiki-en`. The available options for 
`--dataset` for OSCAR data include `oscar-sin`, `oscar-eng`, `oscar-pol`, `oscar-fas`.
data includes

The files are UTF-8 encoded text files with one wiki article per line.

## Development

This repo is for transforming the given datasets into the appropriate form and going the the process of:

1. Download from original source
2. Any manipulation of internal form
3. Upload to the destinations

### Current Functionality

Currently we are going down the path of having dataset prep be consistent for task types. The example dataset to build out this proof of concept is using the MNIST dataset. The idea is to have each dataset implement a `Processer` that is responsible for implementing a `download`, `process`, and `transfer` method. The `Processer` inherits from the `BaseProcesser` class that contains general purpose functionality in helping transform source datasets into the LwLL format.

The format is looking like we will keep a consistent folder schema throughout task types, but it may make sense in the future to in addition to the main `BaseProcesser` class, to also have "original format" Processers. i.e. Object classification original source that is arranged in class / folder structure vs. object classification original source arranged in label file format. This is not super important for the moment, but something to consider as making PR's to this code base.

Because we adhere to the interface of the `BaseProcesser` class, we will be able to write a main script that calls these methods uniformly. An example of going through data collection to processing of the example MNIST dataset is as follows:

```python
from datasets.mnist.main import MNISTProcesser

proc = MNISTProcesser()
proc.download()
proc.process()

```

The output files are transformed to the desired folder schema with consistent random seed based sampling in the users home directory under `~/lwll_datasets`.

### Dataset Metadata Schema

The dataset metadata is put into 3 different locations for proper consumption by appropriate resources.

#### Within Directory Structure

We store the base images within the `datasets/lwll_datasets` directory, these are what are included in the compressed files for reference. These do not include labels because those are expected to be queried through the api.

Schema for `image_classification` and `object_detection` tasks:

```sh
-- datasets
-- -- mnist
-- -- -- mnist_sample
-- -- -- -- train
-- -- -- -- -- img1.png
-- -- -- -- -- img2.png
-- -- -- -- -- etc...
-- -- -- -- test
-- -- -- -- -- img1000.png
-- -- -- -- -- img1001.png
-- -- -- -- -- etc...
-- -- -- mnist_full (Same as sample version but full dataset in this directory)
-- -- -- -- etc...
```

Schema for `machine_translation` tasks:

```sh
-- datsets
-- -- europarl-da-en
-- -- -- europarl-da-en_sample
-- -- -- -- train_data.feather
-- -- -- -- test_data.feather
-- -- -- europarl-da-en_full
-- -- -- -- train_data.feather
-- -- -- -- test_data.feather
```

## Task Type Standardized Label Schemas

We standardize the label format for consumption by the api for querying and delivering labels to the TA1 systems. The label file is standardized by task type as such (\*Note that the examples say 'class1', the actual descriptive lables are present, we do not obfuscate the actual labels in any way):

### Image Classification

```csv
id, label
'img_1.png', 'class3'
'img_2.png', 'class4'
'img_3.png', 'class2'
...
etc...
```

### Object Detection

```csv
id, label, class
'img_1.png', '<x_min>, <y_mmin>, <x_max>, <y_max>', 'class1'
'img_1.png', '<x_min>, <y_mmin>, <x_max>, <y_max>', 'class1'
'img_2.png', '<x_min>, <y_mmin>, <x_max>, <y_max>', 'class5'
'img_2.png', '<x_min>, <y_mmin>, <x_max>, <y_max>', 'class2'
'img_2.png', '<x_min>, <y_mmin>, <x_max>, <y_max>', 'class1'

```

### Machine Translation

```csv
TBD...
```


HISTORY.md
## August 2020 Eval

This history file was started 6/22/2020, thus does not contain details of items that were added prior to that date.

### Enhancements

-   Added imagenet_bbox [#45](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/45) [!21](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/21)
-   Added a `CONTRIBUTING.md` file for instructions on how to contribute to this repository [!28](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/28)
-   Restricted test sets for machine translation to 2000 senstances [lwll_api#82](https://gitlab.lollllz.com/lwll/lwll_api/-/issues/82)
-   Added google_open_image [#26](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/26) [!29](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/29)
-   Added xview [#46](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/46) [!35](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/35)
-   Added version check functionality for download scripts [!33](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/33)
-   Added validation logic [#34](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/34) and secondary seed label [#33](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/33) [!38](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/38)
-	Added validation and reprocess image classification and object detection datasets [#64](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/62) [!41](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/41)
-   Added stanford campus dataset [#53](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/53) [!45](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/45)
-	Added VisDrone [#52](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/52) [!42](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/42)
-   Added Deep Fashion 2 [#56](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/56) [!44](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/44)
-   Added Deep Fashion [#55](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/55) [!48](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/48)
-   Added HMDB [#50](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/50) [!50](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/50)
-	Added Products10k [#64](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/64) [!46](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/46)
-   Added UCF101 [#51](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/51) [!53](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/53)
-   Added MSL Curiosity Images[#78](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/78) [!58](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/58)
-   Added Mars Surface Images[#77](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/77) [!59](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/59)
-   Added Domain Net Zero Shot Learning [#86](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/86) [!61](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/61)
-   Added Wikimatrix [#79](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/79), [#80](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/80), and [#81](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/81) [!60](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/60)
-	Added Widerperson [#69](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/69) [!66](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/66)
-   Added Cifar100 Zero Shot Learning [#87](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/87), [!69](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/69)
-   Added CC Aligned [#83](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/83), [#84](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/84), [#82](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/82), [!65](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/65)
-	Added Pool Car dataset [#47](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/47), [!23](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/23)
-	Added ImageNet with bounding box [#45](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/45), [!21](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/21)
-   Added 'check' flag to download script [#88](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/88), [!76](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/76)
-   Added ImageNet_sketch dataset [#73](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/73), [!86](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/86)
-   Updated download_monolingual.py to work with OSCAR datasets [#111](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/111)
-   Added quickdraw dataset [#74](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/74), [!102](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/102)
-   Added feathersv1 dataset [#58](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/58), [!103](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/103)
-   Added zero shot learning to mnist [!105](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/105)
-	Added iNaturalist dataset [#57](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/57), [!106](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/106)
-	Added Imagenet-22k dataset [#94](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/94), [!108](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/108)
-   Added Chest Xray dataset [#75](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/75), [!109](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/109)
-	Added UCMerced LandUse dataset [#114](https://gitlab.lollllz.com/lwll/dataset_prep/-/issues/114), [!113](https://gitlab.lollllz.com/lwll/dataset_prep/-/merge_requests/113)

lwll_dataset_prep/xview/main.py
# Copyright (c) 2023 California Institute of Technology (“Caltech”). U.S.
# Government sponsorship acknowledged.
# All rights reserved.
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions are
# met:
#
# * Redistributions of source code must retain the above copyright notice, this
#   list of conditions and the following disclaimer.
# * Redistributions in binary form must reproduce the above copyright notice,
#   this list of conditions and the following disclaimer in the documentation
#   and/or other materials provided with the distribution.
# * Neither the name of Caltech nor its operating division, the Jet Propulsion
#   Laboratory, nor the names of its contributors may be used to endorse or
#   promote products derived from this software without specific prior written
#   permission.
#
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
# AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT OWNER OR CONTRIBUTORS BE
# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
# POSSIBILITY OF SUCH DAMAGE.

from lwll_dataset_prep.dataset_scripts.schema import DatasetDoc
from lwll_dataset_prep.dataset_scripts.process_interface import BaseProcesser
from dataclasses import dataclass, field
from typing import List, Dict
from pathlib import Path
import pandas as pd
import numpy as np
import shutil
from lwll_dataset_prep.logger import log
import os
import cv2
import glob
import json
import yaml


@dataclass
class xview(BaseProcesser):
    """
    Number of images: 1413

    classes: 60 (e.g. aircraft hangar, passenger car, shipping container)
    training images: 847 (600345 objects)
    validation images: 282 (200291 objects)
    testing images: 284

    WARNING: The sample dataset does not guarantee to cover all 60 classes

    The source data is organized as follows:

        xview
        ├── train_images_small
        │   ├── 1.tif
        │   ├── 2.tif
        │   .
        │   .
        │   .
        │   └── N.tif
        ├── val_images_small
        │   ├── 100.tif
        │   ├── 101.tif
        │   └── 102.tif
        │   .
        │   .
        │   .
        │   └── M.tif
        └── xView_train.geojson
    """

    data_path: Path = Path('lwll_datasets')
    labels_path: Path = Path('lwll_labels')
    tar_path: Path = Path('lwll_compressed_datasets')
    _path_name: str = 'xview'

    _urls: List[str] = field(default_factory=lambda:
                             ['https://data.kitware.com/api/v1/item/5f36ecb39014a6d84e7757ef/download',
                              'https://data.kitware.com/api/v1/item/5f36dfc49014a6d84e7739e1/download'])

    _task_type: str = 'object_detection'

    # Numbers are a guess
    _sample_size_train: int = 100
    _sample_size_test: int = 60
    _k_seed: List[int] = field(default_factory=lambda: [1])

    _valid_extensions: List[str] = field(default_factory=lambda: ['.tif'])

    def __post_init__(self) -> None:
        self.path: Path = self.data_path.joinpath(self._path_name)

        self._fnames: List[str] = [
            'train_images.tgz',
            'train_labels.tgz',
        ]

        self.full_path = str(self.data_path.joinpath(
            self._path_name).joinpath(f'{self._path_name}_full/train'))
        self.sample_path = str(self.data_path.joinpath(
            self._path_name).joinpath(f'{self._path_name}_sample/train'))

    def is_faulty(self, img: np.ndarray, thrs: int = 20) -> bool:
        # checks if image has 'thrs' black columns
        count = 0
        for x in range(img.shape[1]):
            sum_cols = 0
            for y in range(img.shape[0]):
                sum_cols += img[y][x]
            if sum_cols == 0:
                count += 1

            if count > thrs:
                return True
        return False

    def remove_faulty_imgs(self) -> List[str]:
        removed = []
        img_path = os.path.join(self.path, 'train_images')
        img_list = os.listdir(img_path)
        for im in img_list:
            img_gray = cv2.cvtColor(cv2.imread(os.path.join(img_path, im)), cv2.COLOR_BGR2GRAY)
            faulty = self.is_faulty(img_gray)
            if (faulty):
                os.remove(os.path.join(img_path, im))
                removed.append(im)
                log.info(f'Faulty image removed: {im}')
                # shutil.move(os.path.join(img_path, im), os.path.join(remove_path, im))
        return removed

    def download(self) -> None:
        # Download
        for url, fname in zip(self._urls, self._fnames):
            self.download_data_from_url(url=url, dir_name=self._path_name, file_name=fname, overwrite=True)
            log.info("Done")

    def process(self) -> None:
        log.info('Processing xview dataset')

        # Extract the tar
        for fname in self._fnames:
            self.extract_tar(dir_name=self._path_name, fname=fname)

        cur_dir_path = Path(os.path.dirname(os.path.realpath(__file__)))
        with open(cur_dir_path / 'class_mappings.yaml', 'r') as file:
            self._class_ids_map: Dict[int, str] = yaml.load(file)

        # Create our output directories
        self.setup_folder_structure(dir_name=self._path_name)

        # Removing annoying hidden files
        for fname in glob.glob(f'{self.path}/train_images/.*'):
            os.remove(fname)

        log.info('Removing faulty images from dataset')
        removed = self.remove_faulty_imgs()
        # removed = ['2004.tif', '1450.tif', '1863.tif', '1081.tif' , '2012.tif' , '100.tif', '2599.tif', '2000.tif','1870.tif', \
        # '1988.tif', '2306.tif', '1237.tif', '1154.tif', '89.tif', '2314.tif', '2513.tif', '84.tif', '90.tif', '1429.tif', '2530.tif', \
        # '1372.tif', '95.tif', '1164.tif', '2309.tif', '83.tif', '752.tif', '1459.tif', '1465.tif', '481.tif', '2542.tif', '1739.tif', \
        # '1880.tif', '1072.tif', '2591.tif', '1891.tif']

        # get labels
        log.info('Getting labels')
        with open(self.path / f'xView_train.geojson', 'r') as file:
            labels = json.load(file)['features']

        # convert ann to dict
        anns = list()
        for lab in labels:
            class_label = lab['properties']['type_id']
            id_label = lab['properties']['image_id']
            bbox_label = lab['properties']['bounds_imcoords']
            # Skipping some class labels https://github.com/DIUx-xView/xView1_baseline/issues/3
            if class_label in [75, 82] or id_label == '1395.tif':
                continue
            anns.append(dict({
                'id': id_label,
                'bbox': bbox_label,
                'class': self._map_class_id_to_name(class_label)
            }))

        # convert to pandas
        anns_df: pd.DataFrame = pd.DataFrame(anns)
        anns_df = anns_df.sample(frac=1, random_state=34).reset_index(drop=True)

        # Get Classes
        classes = anns_df['class'].unique().tolist()
        image_fns = list(anns_df['id'].unique().tolist())
        # Split dataset by image name
        # TODO: Use interger programming method:
        #  https://github.com/bmreiniger/datascience.stackexchange/blob/master/54450.ipynb
        n_train = int(len(image_fns) * 0.7)
        train_fns = image_fns[:n_train]
        test_fns = image_fns[n_train:]
        # Manually pull out 107.tif from train to test to enforce 60 classes in test
        train_fns.remove('107.tif')
        test_fns.append('107.tif')

        df_train = anns_df[anns_df['id'].isin(train_fns)]
        df_test = anns_df[anns_df['id'].isin(test_fns)]

        # removing filtered images
        df_train = anns_df[~anns_df['id'].isin(removed)]
        df_test = anns_df[~anns_df['id'].isin(removed)]

        log.info(f'Train has {len(df_train["class"].unique())} classes and '
                 f'Test has {len(df_test["class"].unique())} classes')

        train_full_fns = self.get_full_fname_unqiue_paths(df_train)
        test_full_fns = self.get_full_fname_unqiue_paths(df_test)

        self.original_paths_to_destination(dir_name=self._path_name, orig_paths=train_full_fns,
                                           dest_type='train', delete_original=False)
        self.original_paths_to_destination(dir_name=self._path_name, orig_paths=test_full_fns,
                                           dest_type='test', delete_original=False)

        # Create our sample subsets
        # Per performer's request, generating samples exactly the same as train/val data
        df_train_sample, df_test_sample = self.create_label_files(dir_name=self._path_name,
                                                                  df_train=df_train,
                                                                  df_test=df_test,
                                                                  samples_train=len(df_train),
                                                                  samples_test=len(df_test),
                                                                  many_to_one=True)

        # Copy appropriate sample data to sample location
        self.copy_from_full_to_sample_destination(dir_name=self._path_name,
                                                  df_train_sample=df_train_sample,
                                                  df_test_sample=df_test_sample)

        # Create our `dataset.json` metadata file
        dataset_doc = DatasetDoc(name=self._path_name,
                                 dataset_type='object_detection',
                                 sample_number_of_samples_train=len(df_train_sample["id"].unique()),
                                 sample_number_of_samples_test=len(df_test_sample["id"].unique()),
                                 sample_number_of_classes=len(classes),
                                 full_number_of_samples_train=len(df_train["id"].unique()),
                                 full_number_of_samples_test=len(df_test["id"].unique()),
                                 full_number_of_classes=len(classes),
                                 number_of_channels=3,
                                 classes=classes,
                                 language_from=None,
                                 language_to=None,
                                 sample_total_codecs=None,
                                 full_total_codecs=None,
                                 license_link='https://challenge.xviewdataset.org/data-download',
                                 license_requirements='agree_to_termsLicense Link at '
                                                      'https://challenge.xviewdataset.org/data-download',
                                 license_citation='N/A',  # noqa
                                 )
        self.save_dataset_metadata(dir_name=self._path_name, metadata=dataset_doc)

        # Validating sample datasets
        log.info("Validating sample datasets")
        self.validate_output_structure(name=self._path_name,
                                       orig_train_paths=train_full_fns,
                                       orig_test_paths=test_full_fns,
                                       classes=classes,
                                       train_df=df_train,
                                       test_df=df_test,
                                       sample_train_df=df_train_sample,
                                       sample_test_df=df_test_sample,
                                       metadata=dataset_doc,
                                       many_to_one=True)

        # Cleanup tar extraction intermediate
        log.info("Cleaning up extracted tar copy..")
        for f in self._fnames:
            # We assume the tar files have no '.'s in their name before `.tar.gz` or just `.tar`
            d = f.split('.')[0]
            if d == 'train_labels':
                os.unlink(self.path.joinpath('xView_train.geojson'))
                os.unlink(self.path.joinpath(f))
            else:
                shutil.rmtree(self.path.joinpath(d))
                os.unlink(self.path.joinpath(f))

        log.info("Done")

    def transfer(self) -> None:
        log.info("Pushing artifacts to appropriate cloud resources...")
        self.push_data_to_cloud(dir_name=self._path_name,
                                dataset_type='development',
                                task_type=self._task_type)
        log.info("Done")

    def _map_class_id_to_name(self, class_id: int) -> str:
        """
        Args:
            class_id: number for the class

        Return:
            class name
        """
        return self._class_ids_map[class_id]

    def get_full_fname_unqiue_paths(self, anns: pd.DataFrame) -> np.array:
        """ Add full path to the image filenanes in ID.

        """
        return list(map(lambda x: Path().joinpath(f'{self.path}/train_images/', x),
                        sorted(anns['id'].unique().tolist())))
